api/v1alpha1/linodecluster_webhook.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2023 Akamai Technologies, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"slices"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	. "github.com/linode/cluster-api-provider-linode/clients"
+)
+
+// log is for logging in this package.
+var linodeclusterlog = logf.Log.WithName("linodecluster-resource")
+
+// SetupWebhookWithManager will setup the manager to manage the webhooks
+func (r *LinodeCluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// TODO(user): EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable updation and deletion validation.
+//+kubebuilder:webhook:path=/validate-infrastructure-cluster-x-k8s-io-v1alpha1-linodecluster,mutating=false,failurePolicy=fail,sideEffects=None,groups=infrastructure.cluster.x-k8s.io,resources=linodeclusters,verbs=create,versions=v1alpha1,name=vlinodecluster.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &LinodeCluster{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *LinodeCluster) ValidateCreate() (admission.Warnings, error) {
+	linodeclusterlog.Info("validate create", "name", r.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultWebhookTimeout)
+	defer cancel()
+
+	return nil, r.validateLinodeCluster(ctx, &defaultLinodeClient)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *LinodeCluster) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
```

Comment:
can we add some tests for each of these validate functions? even if it is just confirming they are currently doing nothing?

Reply:
Unfortunately not. 

1. These `Validate*` methods are functionally similar to the individual `Reconcile` methods - they are _only_ intended to be called by the Controller's Manager. In our case, they're also intended to inject real dependencies (e.g. `linodego.Client` and cannot be mocked.

2. For the methods that do nothing (e.g. `ValidateUpdate`, `ValidateDelete`), they do not get called since the registration of the underlying operations into Kubernetes is configured through a [controller-gen marker](https://github.com/linode/cluster-api-provider-linode/pull/320/files#diff-4df2d4fcaeca43c236826f56e655e92482e2cf05ed11b79c08546e6440bbff77R47-R48). These no op methods are currently only scaffolding functionally. I don't think it would be particularly valuable to add a unit test here.